### PR TITLE
fix: quoted identifiers for source names

### LINKED ISSUE #1466: KSQL should support object names with the - character (hyphen, dash)
`-` is rejected but `_` is valid. At minimum, we should catch and report this limitation better to the user, as the current message is very unclear: 

Doesn't work: 
```
ksql> CREATE STREAM ratings-with-customer-data WITH (PARTITIONS=1) AS \
> SELECT R.RATING_ID, R.CHANNEL, R.STARS, R.MESSAGE, \
>        C.ID, C.CLUB_STATUS, C.EMAIL, \
>        C.FIRST_NAME,  C.LAST_NAME \
> FROM RATINGS R \
>      LEFT JOIN CUSTOMERS C \
>        ON R.USER_ID = C.ID \
> WHERE C.FIRST_NAME IS NOT NULL ;
line 1:22: mismatched input '-' expecting ';'
Caused by: org.antlr.v4.runtime.InputMismatchException
```

Works:
```
ksql> CREATE STREAM ratings_with_customer_data WITH (PARTITIONS=1) AS \
> SELECT R.RATING_ID, R.CHANNEL, R.STARS, R.MESSAGE, \
>        C.ID, C.CLUB_STATUS, C.EMAIL, \
>        C.FIRST_NAME, C.LAST_NAME \
> FROM RATINGS R \
>      LEFT JOIN CUSTOMERS C \
>        ON R.USER_ID = C.ID \
> WHERE C.FIRST_NAME IS NOT NULL ;

 Message
----------------------------
 Stream created and running
----------------------------
```

### LINKED ISSUE #2171: Cannot DROP or TERMINATE when names of streams, tables, queries are quoted (backticks)
```show streams;

Stream Name                 | Kafka Topic                 | Format 
--------------------------------------------------------------------
ks.hgame.projects.11_stream | ks.hgame.projects.11_stream | JSON   
vs.hgame.projects.1_stream  | maxwell.hgame.projects      | JSON   
vs.hgame.projects.n1_stream | maxwell.hgame.projects      | JSON   
--------------------------------------------------------------------
ksql> drop stream `vs.hgame.projects.1_stream`;
Cannot drop vs.hgame.projects.1_stream. 
The following queries read from this source: [CSAS_ks.hgame.projects.11_stream_3]. 
The following queries write into this source: []. 
You need to terminate them before dropping vs.hgame.projects.1_stream.
ksql> terminate `CSAS_ks.hgame.projects.11_stream_3`;
java.lang.Exception: No running query with id `CSAS_ks.hgame.projects.11_stream_3` was found
ksql> show queries;

Query ID                           | Kafka Topic                 | Query String                                                                                                                                                                                                                                                                                                                                                                                                                                                 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 CSAS_ks.hgame.projects.11_stream_3 | ks.hgame.projects.11_stream | CREATE STREAM `ks.hgame.projects.11_stream` WITH (VALUE_FORMAT='JSON') AS SELECT data->`userid` AS userid, data->`iscancel` AS iscancel, data->`totalprice` AS totalprice, data->`bouns` AS bouns, data->`bonustime` AS bonustime, data->`writetime` AS writetime, data->`deducttime` AS deducttime, data->`projectid` AS projectid, data->`lotteryid` AS lotteryid, data->`methodid` AS methodid, data->`modes` AS modes FROM `vs.hgame.projects.1_stream`; 
---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
For detailed information on a Query run: EXPLAIN <Query ID>;
ksql> terminate `CSAS_ks.hgame.projects.11_stream_3`;
java.lang.Exception: No running query with id `CSAS_ks.hgame.projects.11_stream_3` was found
ksql> 
```

### LINKED ISSUE #2176: Invalid Expression table.ROWTIME when table was created with backticks around name
```
ksql> CREATE TABLE `ks_hgame_projects_n111_table` (userid BIGINT, iscancel INTEGER, tprice DOUBLE, bonus DOUBLE, btimestamp BIGINT, wtimestamp BIGINT, dtimestamp BIGINT, pid BIGINT, lid INTEGER, mid BIGINT, modes INTEGER) WITH (kafka_topic='ks_hgame_projects_n11_stream', KEY='userid', VALUE_FORMAT='JSON');

Message       
---------------
Table created 
---------------
ksql> show tables;

Table Name                   | Kafka Topic                  | Format | Windowed 
---------------------------------------------------------------------------------
ks_hgame_projects_n111_table | ks_hgame_projects_n11_stream | JSON   | false    
---------------------------------------------------------------------------------
ksql> select * from `ks_hgame_projects_n111_table`;
Invalid Expression ks_hgame_projects_n111_table.ROWTIME.
ksql> select p.userid from `ks_hgame_projects_n111_table` p;
1008
6219
6219
^CQuery terminated
ksql> select p.ROWTIME from `ks_hgame_projects_n111_table` p;
1542766804803
^CQuery terminated
ksql> select * from `ks_hgame_projects_n111_table`;
Invalid Expression ks_hgame_projects_n111_table.ROWTIME.
ksql>
```

### PR DESCRIPTION
**Pull request by agavra**
Fixes #1466 
Fixes #2171 
Fixes #2176  

### Description 

Fixes two of the long-running issues around quoted identifiers and source names:
- `AstBuilder` and `DataSourceExtractor` no longer force upper case even when the source name is quoted
- `ParserUtil` enforces that source names match state store requirements (topic naming convention)
- `sourceName` is used directly in our grammar to give strong typing in the code

### Testing done 

- went through a ton of tickets and manually tried reproducing them, fixed them as they went
- QTT tests
- Unit tests

```sql
ksql> CREATE STREAM `foo-bar` (id VARCHAR) WITH (kafka_topic='foo', value_format='JSON', partitions=1);

 Message
----------------
 Stream created
----------------
ksql> INSERT INTO `foo-bar` (id) VALUES ('123');
ksql> CREATE STREAM `foo-too` AS SELECT * FROM `foo-bar`;

 Message
------------------------------------------------------------------------------------
 Stream foo-too created and running. Created by query with query ID: CSAS_foo-too_5
------------------------------------------------------------------------------------
ksql> SELECT * FROM `foo-too` EMIT CHANGES;
+------------------------------+------------------------------+------------------------------+
|ROWTIME                       |ROWKEY                        |ID                            |
+------------------------------+------------------------------+------------------------------+
|1572362449405                 |null                          |123                           |
|1572362449405                 |null                          |123                           |
^CQuery terminated
ksql> TERMINATE `CSAS_foo-too_5`;

 Message
-------------------
 Query terminated.
-------------------
ksql> DROP STREAM `foo-too`;

 Message
------------------------------------------------
 Source `foo-too` (topic: foo-too) was dropped.
------------------------------------------------
```

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java [hunk 1]
```diff
@@ -458,6 +459,29 @@ public void testReservedRowKeyAlias() {
...
+    expectedException.expectMessage("Got: 'foo!bar'");
+
+    // When:
+    KsqlParserTestUtil.buildSingleAst(
+        "CREATE STREAM `foo!bar` WITH (kafka_topic='foo', value_format='AVRO');",
+        metaStore);
+  }
+
+  @Test
+  public void shouldAllowEscapedTerminateQuery() {
```
Comment (vcrfxia) [inline ksql-parser/src/test/java/io/confluent/ksql/parser/KsqlParserTest.java]: Should we also add a test for `INSERT INTO` with a quoted source name, either here or as a QTT?

**Comment by apurvam** [comment]
great to see these patches.. they are fixing long standing issues with KSQL which have proven super frustrating!

My knowledge of this code base is dated, but I'm trying to understand how these changes fix the issues where quoted source names are not handled correctly. Is it primarily the switch to use `SourceName` consistently everywhere in the code base that fixes these inconsistencies?

**Comment by agavra** [comment]
> My knowledge of this code base is dated, but I'm trying to understand how these changes fix the issues where quoted source names are not handled correctly. Is it primarily the switch to use SourceName consistently everywhere in the code base that fixes these inconsistencies?

@apurvam We've just made changes across the board to make sure that the proper names are being piped through the codebase. This specific PR helps (1) quoted source names that were lowercase because previously we forced uppercase even if it was quoted and (2) use the correct parsing from TERMINATE query. The main changes (i.e. around java code generation) were done previously. The `SourceName` change makes things easier to reason about, but I'm not sure it helped "fix" any issues.

**Comment by apurvam** [comment]
Got it, thanks @agavra !

**Comment by pkgonan** [comment]
Is there any solution about Describe command ?

[Request]
```
DESCRIBE 'abc-data.kk_bb.ee_dd.mm_zz' EXTENDED;
```

[Response]
```
ksql> DESCRIBE 'abc-data.kk_bb.ee_dd.mm_zz' EXTENDED;
line 1:10: no viable alternative at input 'DESCRIBE 'abc-data.kk_bb.ee_dd.mm_zz''
Statement: DESCRIBE 'abc-data.kk_bb.ee_dd.mm_zz' EXTENDED;
Caused by: line 1:10: no viable alternative at input 'DESCRIBE
	'abc-data.kk_bb.ee_dd.mm_zz''
Caused by: org.antlr.v4.runtime.NoViableAltException
```

**Comment by vcrfxia** [comment]
Hi @pkgonan have you tried using backticks, rather than single quotes? As in:
```
DESCRIBE `abc-data.kk_bb.ee_dd.mm_zz` EXTENDED;
```
instead of `'abc-data.kk_bb.ee_dd.mm_zz'`?

**Comment by agavra** [comment]
I tried repro'ing locally and I noticed a few things:

1. you can't create an entity with `.` in it's name, so I"m wondering how you got the query you're trying to describe:
```
ksql> create stream `b.b.q` as select * from foo;
Illegal argument at Line: 1, Col: 15. Source names may only contain alphanumeric values, '_' or '-'. Got: 'b.b.q'
```

2. You can't describe, queries, just tables or streams:
```
ksql> describe `CSAS_B-B-Q_3`;
Could not find STREAM/TABLE 'CSAS_B-B-Q_3' in the Metastore
```

3. I can't repro the issue using only `-`:
```
ksql> describe `b-b-q` extended;

Name                 : b-b-q
Type                 : STREAM
Timestamp field      : Not set - using <ROWTIME>
Key format           : JSON
...
```

**Pull request merged**